source/common/router/config_impl.cc

```diff
@@ -305,7 +316,9 @@ RouteEntryImplBase::RouteEntryImplBase(const VirtualHostImpl& vhost,
       path_redirect_(route.redirect().path_redirect()),
       https_redirect_(route.redirect().https_redirect()),
       prefix_rewrite_redirect_(route.redirect().prefix_rewrite()),
-      strip_query_(route.redirect().strip_query()), retry_policy_(route.route()),
+      strip_query_(route.redirect().strip_query()),
+      retry_policy_(!route.route().has_retry_policy() ? RetryPolicyImpl(vhost.config())
```

Comment:
I think it would be cleaner to use a helper function here that accepts both the route and the vhost config that uses the appropriate retry policy to construct the `RetryPolicyImpl`. Parsing the negation and ternary when reading this code it's not super obvious what's going on.

Maybe `RetryPolicyImpl` should just have two ctors: one that accepts the `RetryPolicy` and one that accepts no args that uses the default retry behavior. That way all the logic for deciding what retry policy to use can be centralized in the helper function.

Reply:
+1. Looks like we have most specific wins semantics, which is reasonable. Did you consider any kind of composing of the policies, are there situations where that would make sense?

Reply:
I did not consider composing/merging policies... Do we do it for other top level configs? I'll look around, to ensure we are consistent about composing vs most specific wins. 

In our use-case, most specific wins works fine... I _think_ I prefer most specific wins because it's easier to read/reason around... what you see is what you get, as opposed to some derived value from top level config that you forgot.

Also, these `RetryPolicy` messages are not that big, so the copy/pasta required when explicitly defining a policy for each route that differs from the vhost policy would still be way less than what we have now...

Reply:
+1 on most specific wins, it makes it easier to express configurations where the per route policy is less aggressive. For instance, if you want to retry on `conflict,5xx` on the vhost level but only `5xx` on the route level.